### PR TITLE
Use new FFT operators in stft

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes https://github.com/pytorch/pytorch/issues/42175#issuecomment-719933913

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46912 Optimize torch.fft.fftn for real input with an r2c transform
* #49022 stft: Change require_complex warning to an error
* #48905 Remove unused operator at::_fft_with_size
* **#47601 Use new FFT operators in stft**
* #46911 Improve torch.fft n-dimensional transforms

Differential Revision: [D25457217](https://our.internmc.facebook.com/intern/diff/D25457217)

### CODE REVIEW

File: aten/src/ATen/native/SpectralOps.cpp [hunk 1]
```diff
@@ -724,12 +725,26 @@ Tensor stft(const Tensor& self, const int64_t n_fft, const optional<int64_t> hop
...
   if (return_complex) {
-    return at::view_as_complex(out);
-  } else {
     return out;
+  } else {
+    return at::view_as_real(out);
   }
 }
 
+// Like view_as_complex but may clone if the input can't be directly viewed as complex
```
Comment (mruberry) [inline aten/src/ATen/native/SpectralOps.cpp:734]: Can you elaborate on how you expect this function to be used?

It's a little odd to me that it's so aggressive at coercing tensors to complex, but I guess that makes sense for the current state deprecated state of stft. If that's the case, however, the comment should probably clarify.

Is this the right PR to update stft to the next stage in the deprecation process where it requires `return_complex`?

Comment (peterbell10) [reply aten/src/ATen/native/SpectralOps.cpp:734]: > It's a little odd to me that it's so aggressive at coercing tensors to complex

I don't see the issue. It converts any tensor from the old complex format, to a complex dtype tensor. `view_as_complex` is limited by the fact that it has to create a view of the original tensor but there's no reason to force that limitation onto the user now. Would it be clearer if the comment didn't mention `view_as_complex`?

>Is this the right PR to update stft to the next stage in the deprecation process where it requires `return_complex`? 

That doesn't simplify anything in this PR and I guess could hold up merging, so I would say no. I can open a separate PR for it though.

>Also, should we deprecate istft accept a float tensor mimicking a half tensor now? Then after 1.8 we can have stft always return a complex tensor and istft always accept a complex tensor.

Sure, I can do that.

Comment (mruberry) [reply aten/src/ATen/native/SpectralOps.cpp:734]: A separate PR for those latter issues sounds good.

As for this particular function, mentioned view_as_complex in the comment seems fine. Maybe just making the comment more specific by saying that this function is only intended to be used for istft and picking a name for it that matches? That is, I would think that once istft is updated to ONLY accept complex inputs this function will be killed, right?

Comment (peterbell10) [reply aten/src/ATen/native/SpectralOps.cpp]: @mruberry I've reworded the comment, PTAL.

**Comment by mruberry** [comment]
Hey @peterbell10! Is this ready for another review?

**Review by mruberry — APPROVED** [review]
Cool! Thanks @peterbell10!

**Pull request closed without merging**